add top-level example for binary heap

wilzbach opened the pull request:
- adds a top-level example for the binary heap
- adds support to extend/grow a normal array (didn't see a reason why this shouldn't be allowed)
- removes the out-commented private (seems like the bug was fixed)

Ping @schveiguy

schveiguy commented:
I need to fix the authorship on these submodules :) I only wrote the RedBlackTree, and Andrei and others wrote everything else!

burner commented:
@wilzbach can you split these kind of PR's either it is a doc PR or it is an code enhancement.
and comments like "seems like the bug was fixed" are just not ok. either it is fixed or not.

IMO this PR should be closed and split into multiple new PR's, each with a singular focus.

wilzbach commented:
> @wilzbach can you split these kind of PR's either it is a doc PR or it is an code enhancement.
and comments like "seems like the bug was fixed" are just not ok. either it is fixed or not.

Sorry. Ok, makes sense,

> IMO this PR should be closed and split into multiple new PR's, each with a singular focus.

I rebased this PR to be just about the added unittest example to the documentation. Hope this reduces the noise a bit ;-)

burner commented on std/container/binaryheap.d:
```diff
@@ -25,6 +25,24 @@ import std.traits;
+///
+unittest
+{
+    import std.algorithm : equal;
```
isn't equal in std.algorithm.comparison ? I mean if you use selected imports, use them on deepest nested module.

burner commented:
Other than my remark about the imports LGTM

wilzbach commented:
> Other than my remark about the imports LGTM

Rebased, but now it fails, because it depends on the bug fix #4359 :/
-> I removed the insert part

```
    auto q = heapify!"a > b"([1, 4, 5]);
    q.insert(2);
    assert(q.front == 1);
    q.insert(0);
    assert(q.front == 0);
``

and will the include it with #4359?

burner commented:
Auto-merge toggled on
